src/rgw/rgw_sal_motr.cc

```diff
@@ -199,9 +199,8 @@ int static update_bucket_stats(const DoutPrefixProvider *dpp, MotrStore *store,
   bkt_stat.total_size += multiplier * size;
   bkt_stat.actual_size += multiplier * actual_size;
 
-  bl.clear();
-  bkt_header.encode(bl);
-  rc = store->do_idx_op_by_name(user_stats_iname, M0_IC_PUT, bucket_name, bl);
+  bkt_header.encode(update_bl);
```

Comment:
Strange. So does it crash on bl.clear() or encode(bl)? Any idea why?

Reply:
we can try setting the constant iterator to null, before clearing the bufferlist. Maybe that might be causing an issue.

Reply:
It would probably caused by encode() as we are trying to use the bl which is already disposed.

>>we can try setting the constant iterator to null, before clearing the bufferlist. Maybe that might be causing an issue. clearing a bufferlist is important before encoding again.
I feel an additional bufferlist would be safer to hold the updated stats. Not too much of overhead.

Reply:
```c++
    SnapPayload payload;
    // clear the bufferlist that may have been populated by the call
    // to _posix_acl_create(). MDS mksnap does not make use of it.
    // So, reuse it to pass metadata payload.
    bl.clear();
    payload.metadata = metadata;
    encode(payload, bl);
```
ceph code is reusing the bufferlist, so there must be something wrong we are doing. Once we understand the bufferlist it will be easier to spot the issue. 
For now, to fix the crash we can use the new bufferlist. But we need to clear these buffers or they may cause the memory leaks. (I am not sure, just guessing).

Reply:
The way I understand bufferlist is, it maintains reference count for all the memory blocks internally allocated, these memory blocks are managed using linked list. As soon as the reference count goes to zero for the memory block, it gets de-allocated. 
There is much complex logic of managing and sharing these memory blocks within and may be among different bufferlist (not too sure on that part).
But yeah we should get more understanding of how it's implemented internally.

Reply:
>  As soon as the reference count goes to zero for the memory block, it gets de-allocated.

root node does not seem to be deleted on clear() call. It is always keeping the _root node and begin() also starts the list from _root.next. 
```c++
      iterator begin() {
	return _root.next;
      }

      void clear_and_dispose() {
	for (auto it = begin(); it != end(); /* nop */) {
	  auto& node = *it;
	  it = it->next;
	  ptr_node::disposer()(&node);
	}
	_root.next = &_root;
	_tail = &_root;
      }
```